### Import necessary libraries


In [216]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.metrics.pairwise import cosine_similarity
import pickle

### Load the datasets into a pandas DataFrame and Check for Basics stuffs

In [217]:
books = pd.read_csv(r'D:\ML Projects\Recommender System for Books\Data\Books.csv')
users = pd.read_csv(r'D:\ML Projects\Recommender System for Books\Data\Users.csv')
ratings = pd.read_csv(r'D:\ML Projects\Recommender System for Books\Data\Ratings.csv')

C:\Users\RAVI TEJ\AppData\Local\Temp\ipykernel_13336\2719581062.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv(r'D:\ML Projects\Recommender System for Books\Data\Books.csv')


In [218]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [219]:
# Display the first few rows of the 'Image-URL-M' column in the books dataset to inspect the URLs
books['Image-URL-M'].head()

0    http://images.amazon.com/images/P/0195153448.0...
1    http://images.amazon.com/images/P/0002005018.0...
2    http://images.amazon.com/images/P/0060973129.0...
3    http://images.amazon.com/images/P/0374157065.0...
4    http://images.amazon.com/images/P/0393045218.0...
Name: Image-URL-M, dtype: object

In [220]:
# Replace 'http' with 'https' in the 'Image-URL-M' column
books['Image-URL-M'] = books['Image-URL-M'].str.replace('http://', 'https://')

# Verify the changes
books['Image-URL-M'].head()

0    https://images.amazon.com/images/P/0195153448....
1    https://images.amazon.com/images/P/0002005018....
2    https://images.amazon.com/images/P/0060973129....
3    https://images.amazon.com/images/P/0374157065....
4    https://images.amazon.com/images/P/0393045218....
Name: Image-URL-M, dtype: object

In [221]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [222]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [223]:
# Print the shape 
print(books.shape)
print(ratings.shape)
print(users.shape)

(271360, 8)
(1149780, 3)
(278858, 3)


In [224]:
# Display concise summary information about the books dataset including the data types
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271358 non-null  object
 3   Year-Of-Publication  271360 non-null  object
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: object(8)
memory usage: 16.6+ MB


In [225]:
# Display concise summary information about the ratings dataset including the data types
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149780 non-null  int64 
 1   ISBN         1149780 non-null  object
 2   Book-Rating  1149780 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 26.3+ MB


In [226]:
# Display concise summary information about the users dataset including the data types
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   278858 non-null  int64  
 1   Location  278858 non-null  object 
 2   Age       168096 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB


In [227]:
# Checking of null values
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [228]:
# Checking of null values
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [229]:
# Checking of null values
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [230]:
# Checking for duplicated rows
books.duplicated().sum()

0

In [231]:
# Checking for duplicated rows
ratings.duplicated().sum()

0

In [232]:
# Checking for duplicated rows
users.duplicated().sum()

0

In [233]:
# Checking for book cover image url that it is cosistent to fetched out or not
books['Image-URL-M'][1]

'https://images.amazon.com/images/P/0002005018.01.MZZZZZZZ.jpg'

# Popularity Based Recommender System


In [234]:
# Merge the ratings dataset with the books dataset based on the 'ISBN' column to include book details in the ratings data
ratings_with_name = ratings.merge(books,on='ISBN')

In [235]:
ratings_with_name

User-ID        ISBN  Book-Rating  \
0         276725  034545104X            0   
1           2313  034545104X            5   
2           6543  034545104X            0   
3           8680  034545104X            5   
4          10314  034545104X            9   
...          ...         ...          ...   
1031131   276688  0517145553            0   
1031132   276688  1575660792            7   
1031133   276690  0590907301            0   
1031134   276704  0679752714            0   
1031135   276704  0806917695            5   

                                                Book-Title      Book-Author  \
0                                     Flesh Tones: A Novel       M. J. Rose   
1                                     Flesh Tones: A Novel       M. J. Rose   
2                                     Flesh Tones: A Novel       M. J. Rose   
3                                     Flesh Tones: A Novel       M. J. Rose   
4                                     Flesh Tones: A Novel       M. J. Rose   
...                                                    ...              ...   
1031131                                    Mostly Harmless    Douglas Adams   
1031132                                        Gray Matter  Shirley Kennett   
1031133  Triplet Trouble and the Class Trip (Triplet Tr...     Debbie Dadey   
1031134  A Desert of Pure Feeling (Vintage Contemporaries)   Judith Freeman   
1031135  Perplexing Lateral Thinking Puzzles: Scholasti...      Paul Sloane   

        Year-Of-Publication                          Publisher  \
0                      2002                   Ballantine Books   
1                      2002                   Ballantine Books   
2                      2002                   Ballantine Books   
3                      2002                   Ballantine Books   
4                      2002                   Ballantine Books   
...                     ...                                ...   
1031131                1995             Random House Value Pub   
1031132                1996  Kensington Publishing Corporation   
1031133                1997                              Apple   
1031134                1997                  Vintage Books USA   
1031135                1997                Sterling Publishing   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-M  \
0        https://images.amazon.com/images/P/034545104X....   
1        https://images.amazon.com/images/P/034545104X....   
2        https://images.amazon.com/images/P/034545104X....   
3        https://images.amazon.com/images/P/034545104X....   
4        https://images.amazon.com/images/P/034545104X....   
...                                                    ...   
1031131  https://images.amazon.com/images/P/0517145553....   
1031132  https://images.amazon.com/images/P/1575660792....   
1031133  https://images.amazon.com/images/P/0590907301....   
1031134  https://images.amazon.com/images/P/0679752714....   
1031135  https://images.amazon.com/images/P/0806917695....   

                                               Image-URL-L  
0        http://images.amazon.com/images/P/034545104X.0...  
1        http://images.amazon.com/images/P/034545104X.0...  
2        http://images.amazon.com/images/P/034545104X

In [236]:
ratings_with_name.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031136 entries, 0 to 1031135
Data columns (total 10 columns):
 #   Column               Non-Null Count    Dtype 
---  ------               --------------    ----- 
 0   User-ID              1031136 non-null  int64 
 1   ISBN                 1031136 non-null  object
 2   Book-Rating          1031136 non-null  int64 
 3   Book-Title           1031136 non-null  object
 4   Book-Author          1031134 non-null  object
 5   Year-Of-Publication  1031136 non-null  object
 6   Publisher            1031134 non-null  object
 7   Image-URL-S          1031136 non-null  object
 8   Image-URL-M          1031136 non-null  object
 9   Image-URL-L          1031132 non-null  object
dtypes: int64(2), object(8)
memory usage: 78.7+ MB


In [237]:
# Group the merged dataset by book title and count the number of ratings for each book
num_rating_df = ratings_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()

# Rename the column 'Book-Rating' to 'num_ratings' to represent the count of ratings
num_rating_df.rename(columns={'Book-Rating':'num_ratings'},inplace=True)
num_rating_df

Book-Title  num_ratings
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.            2
241067                     Ã?Â?rger mit Produkt X. Roman.            4
241068                                Ã?Â?sterlich leben.            1
241069                              Ã?Â?stlich der Berge.            3
241070                                  Ã?Â?thique en toc            2

[241071 rows x 2 columns]

In [238]:
# Calculate the average rating for each book by grouping the merged dataset by book title
avg_rating_df = ratings_with_name.groupby('Book-Title')['Book-Rating'].mean().reset_index()

# Rename the column 'Book-Rating' to 'num_ratings' to represent the count of ratings
avg_rating_df.rename(columns={'Book-Rating': 'avg_rating'}, inplace=True)
avg_rating_df


Book-Title  avg_rating
0        A Light in the Storm: The Civil War Diary of ...    2.250000
1                                   Always Have Popsicles    0.000000
2                    Apple Magic (The Collector's series)    0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...    8.000000
4        Beyond IBM: Leadership Marketing and Finance ...    0.000000
...                                                   ...         ...
241066                                      Ã?Â?lpiraten.    0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.    5.250000
241068                                Ã?Â?sterlich leben.    7.000000
241069                              Ã?Â?stlich der Berge.    2.666667
241070                                  Ã?Â?thique en toc    4.000000

[241071 rows x 2 columns]

In [239]:
# Merge the DataFrames containing the number of ratings and average ratings for each book based on book title
popular_df = num_rating_df.merge(avg_rating_df,on='Book-Title')

popular_df


Book-Title  num_ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241066                                      Ã?Â?lpiraten.            2   
241067                     Ã?Â?rger mit Produkt X. Roman.            4   
241068                                Ã?Â?sterlich leben.            1   
241069                              Ã?Â?stlich der Berge.            3   
241070                                  Ã?Â?thique en toc            2   

        avg_rating  
0         2.250000  
1         0.000000  
2         0.000000  
3         8.000000  
4         0.000000  
...            ...  
241066    0.000000  
241067    5.250000  
241068    7.000000  
241069    2.666667  
241070    4.000000  

[241071 rows x 3 columns]

In [240]:
popular_df = popular_df[popular_df['num_ratings']>=250].sort_values('avg_rating',ascending=False).head(50)
popular_df

Book-Title  num_ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80422        Harry Potter and the Goblet of Fire (Book 4)          387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)          278   
80426   Harry Potter and the Order of the Phoenix (Boo...          347   
80414    Harry Potter and the Chamber of Secrets (Book 2)          556   
191612  The Hobbit : The Enchanting Prelude to The Lor...          281   
187377  The Fellowship of the Ring (The Lord of the Ri...          368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...          575   
211384     The Two Towers (The Lord of the Rings, Part 2)          260   
219741                              To Kill a Mockingbird          510   
183573                                  The Da Vinci Code          898   
187880                 The Five People You Meet in Heaven          430   
180556                             The Catcher in the Rye          449   
196326                          The Lovely Bones: A Novel         1295   
764                                                  1984          284   
144165                           Prodigal Summer: A Novel          253   
128670                                         Neverwhere          265   
206502                            The Secret Life of Bees          774   
168719  Stupid White Men ...and Other Sorry Excuses fo...          283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...          493   
204387                The Red Tent (Bestselling Backlist)          723   
191589               The Hitchhiker's Guide to the Galaxy          268   
129379   Nickel and Dimed: On (Not) Getting By in America          335   
93381                                       Into the Wild          252   
63867                                      Fahrenheit 451          409   
74750                           Girl with a Pearl Earring          526   
136145                                          Outlander          283   
233370  Where the Heart Is (Oprah's Book Club (Paperba...          585   
156102                     Seabiscuit: An American Legend          275   
107962                                         Life of Pi          664   
176845                                     The Bean Trees          389   
2281    A Child Called \It\": One Child's Courage to S...          265   
8434                                       ANGELA'S ASHES          279   
76343                                         Good in Bed          490   
64931   Fast Food Nation: The Dark Side of the All-Ame...          321   
12700                                       American Gods          302   
161645                                 Skipping Christmas          322   
105777  Left Behind: A Novel of the Earth's Last Days ...          318   
189551    The Golden Compass (His Dark Materials, Book 1)          336   
181679                                   The Color Purple          314   
160336                               Silence of the Lambs          256   
8752                                          About a Boy          262   
158138                  Seven Up (A Stephanie Plum Novel)          278   
175097  The Alchemist: A Fable About Following Your Dream          266   
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...          269   
170101                       Suzanne's Diary for Nicholas          457   
111073                                  Lord of the Flies          259   
5664                              A Prayer for Owen Meany          413   
212070   The Vampire Lestat (Vampire Chronicles, Book II)          301   
233851       White Oleander : A Novel (Oprah's Book Club)          356   

        avg_rating  
80434     5.852804  
80422     5.824289  
80441     5.737410  
80426     5.501441  
80414     5.183453  
191612    5.007117  
187377    4.948370  
80445     4.895652  
211384    4.880769  
219741    4.700000  
183573    4.642539  
187880    4.551163  

In [241]:
# Merge the DataFrame containing popular books with the original books dataset based on book title,
# drop duplicate entries based on book title, and select specific columns for the final DataFrame

popular_df = popular_df.merge(books,on='Book-Title').drop_duplicates('Book-Title')[['Book-Title','Book-Author','Year-Of-Publication','Image-URL-M','num_ratings','avg_rating']]
popular_df

Book-Title  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)   
3         Harry Potter and the Goblet of Fire (Book 4)   
5       Harry Potter and the Sorcerer's Stone (Book 1)   
9    Harry Potter and the Order of the Phoenix (Boo...   
13    Harry Potter and the Chamber of Secrets (Book 2)   
16   The Hobbit : The Enchanting Prelude to The Lor...   
17   The Fellowship of the Ring (The Lord of the Ri...   
26   Harry Potter and the Sorcerer's Stone (Harry P...   
28      The Two Towers (The Lord of the Rings, Part 2)   
39                               To Kill a Mockingbird   
47                                   The Da Vinci Code   
53                  The Five People You Meet in Heaven   
55                              The Catcher in the Rye   
62                           The Lovely Bones: A Novel   
63                                                1984   
72                            Prodigal Summer: A Novel   
73                                          Neverwhere   
78                             The Secret Life of Bees   
84   Stupid White Men ...and Other Sorry Excuses fo...   
85   Tuesdays with Morrie: An Old Man, a Young Man,...   
88                 The Red Tent (Bestselling Backlist)   
89                The Hitchhiker's Guide to the Galaxy   
98    Nickel and Dimed: On (Not) Getting By in America   
100                                      Into the Wild   
103                                     Fahrenheit 451   
116                          Girl with a Pearl Earring   
117                                          Outlander   
123  Where the Heart Is (Oprah's Book Club (Paperba...   
124                     Seabiscuit: An American Legend   
126                                         Life of Pi   
130                                     The Bean Trees   
133  A Child Called \It\": One Child's Courage to S...   
135                                     ANGELA'S ASHES   
137                                        Good in Bed   
139  Fast Food Nation: The Dark Side of the All-Ame...   
140                                      American Gods   
141                                 Skipping Christmas   
147  Left Behind: A Novel of the Earth's Last Days ...   
149    The Golden Compass (His Dark Materials, Book 1)   
157                                   The Color Purple   
163                               Silence of the Lambs   
166                                        About a Boy   
173                  Seven Up (A Stephanie Plum Novel)   
175  The Alchemist: A Fable About Following Your Dream   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...   
182                       Suzanne's Diary for Nicholas   
186                                  Lord of the Flies   
192                            A Prayer for Owen Meany   
194   The Vampire Lestat (Vampire Chronicles, Book II)   
195       White Oleander : A Novel (Oprah's Book Club)   

                Book-Author Year-Of-Publication  \
0             J. K. Rowling                1999   
3             J. K. Rowling                2000   
5             J. K. Rowling                1998   
9             J. K. Rowling                2003   
13            J. K. Rowling                2000   
16           J.R.R. TOLKIEN                1986   
17           J.R.R. TOLKIEN                1986   
26            J. K. Rowling                1999   
28           J.R.R. TOLKIEN                1986   
39               Harper Lee                1988   
47                Dan Brown                2003   
53              Mitch Albom                2003   
55            J.D. Salinger                1991   
62             Alice Sebold                2002   
63            George Orwell                1990   
72       Barbara Kingsolver                2001   
73              Neil Gaiman                1998   
78            Sue Monk Kidd                2003   
84            Michael Moore                2002   
85              MITCH ALBOM                1997   
88            

# Collaborative Filtering Based Recommender System


In [242]:
# Group the ratings_with_name dataframe by User-ID and count the number of ratings for each user
# Check if the count is greater than 200 to identify users with substantial rating activity
x = ratings_with_name.groupby('User-ID').count()['Book-Rating'] > 200

# Extract the indices of users who have rated more than 200 books
achhe_users = x[x].index
achhe_users

Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,   6323,
         6543,
       ...
       271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427, 277639,
       278418],
      dtype='int64', name='User-ID', length=811)

In [243]:
# Filter the ratings_with_name dataframe to include only ratings from users with substantial rating activity
filtered_rating = ratings_with_name[ratings_with_name['User-ID'].isin(achhe_users)]
filtered_rating

User-ID        ISBN  Book-Rating  \
2           6543  034545104X            0   
5          23768  034545104X            0   
7          28523  034545104X            0   
15         77940  034545104X            0   
16         81977  034545104X            0   
...          ...         ...          ...   
1030883   275970  1880837927            0   
1030884   275970  188717897X            0   
1030885   275970  1888889047            0   
1030886   275970  1931868123            0   
1030887   275970  3411086211           10   

                                                Book-Title  \
2                                     Flesh Tones: A Novel   
5                                     Flesh Tones: A Novel   
7                                     Flesh Tones: A Novel   
15                                    Flesh Tones: A Novel   
16                                    Flesh Tones: A Novel   
...                                                    ...   
1030883                         The Theology of the Hammer   
1030884  The Ordeal of Integration: Progress and Resent...   
1030885  Pushcart's Complete Rotten Reviews &amp; Rejec...   
1030886  There's a Porcupine in My Outhouse: Misadventu...   
1030887                                         Die Biene.   

                           Book-Author Year-Of-Publication  \
2                           M. J. Rose                2002   
5                           M. J. Rose                2002   
7                           M. J. Rose                2002   
15                          M. J. Rose                2002   
16                          M. J. Rose                2002   
...                                ...                 ...   
1030883                 Millard Fuller                1994   
1030884              Orlando Patterson                1998   
1030885                 Bill Henderson                1998   
1030886                   Mike Tougias                2002   
1030887  Sybil Gr�¤fin Sch�¶nfeldt                1993   

                                    Publisher  \
2                            Ballantine Books   
5                            Ballantine Books   
7                            Ballantine Books   
15                           Ballantine Books   
16                           Ballantine Books   
...                                       ...   
1030883         Smyth &amp; Helwys Publishing   
1030884                Civitas Book Publisher   
1030885                        Pushcart Press   
1030886                    Capital Books (VA)   
1030887  Bibliographisches Institut, Mannheim   

                                               Image-URL-S  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.amazon.com/images/P/1931868123.0...   
1030887  http://images.amazon.com/images/P/3411086211.0...   

                                               Image-URL-M  \
2        https://images.amazon.com/images/P/034545104X....   
5        https://images.amazon.com/images/P/034545104X....   
7        https://images.amazon.com/images/P/034545104X....   
15       https://images.amazon.com/images/P/034545104X....   
16       https://images.amazon.com/images/P/034545104X....   
...                                                    ...   
1030883  https://images.amazon.com/images/P/1880837927....   
1030884  https://images.amazon.com/images/P/188717897X....   
1030885  https://images.amazon.com/images/P/1888889047....   
1030886  https://images.

In [244]:
# Group the filtered ratings dataframe by book title and count the number of ratings for each book
# Check if the count is greater than or equal to 50 to identify books with significant no of ratings
y = filtered_rating.groupby('Book-Title').count()['Book-Rating'] >= 50

# Extract the indices of books that have received at least 50 ratings
famous_books = y[y].index
famous_books

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=706)

In [245]:
# Filter the filtered ratings dataset to include only ratings for books with significant no. of ratings
final_ratings = filtered_rating[filtered_rating['Book-Title'].isin(famous_books)]

# Display the final ratings dataset containing ratings and other details for books having significant no. of ratings
final_ratings

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       https://images.amazon.com/images/P/0446520802....   
65       https://images.amazon.com/images/P/0446520802....   
66       https://images.amazon.com/images/P/0446520802....   
69       https://images.amazon.com/images/P/0446520802....   
74       https://images.amazon.com/images/P/0446520802....   
...                                                    ...   
1026724  https://images.amazon.com/images/P/0531001725....   
1027923  https://images.amazon.com/images/P/0670809381....   
1028777  https://images.amazon.com/images/P/0440910927....   
1029070  https://images.amazon.com/images/P/B0001PIOX4....   
1030868  https://images.amazon.com/images/P/1586210661....   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

#### Vectorization of Books having Users as the elements of vector in higher dimensional space

In [247]:
# Create a pivot table from the final ratings dataset with book titles as rows, User-IDs as columns,
# and Book-Ratings as values to represent user-book interactions
pivot_table = final_ratings.pivot_table(index='Book-Title', columns='User-ID', values='Book-Rating')

# Display the pivot table representing user-book interactions
pivot_table

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [248]:
# Replace missing values (NaNs) in the pivot table with zeros to represent no rating given by a user for a particular book
pivot_table.fillna(0, inplace=True)

In [249]:
# Display the pivot table representing user-book interactions as the a vector in higher dimensional spcae.
# Means Each Rows here a vector no of columns dimensional space
pivot_table

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

##### Here in above pivot table, we Have 706 vectors in the 810 dimensional space.

### Computation of Cosine Similarity (Collaborative Filtering)

In [250]:
from sklearn.metrics.pairwise import cosine_similarity

In [251]:
# Calculate cosine similarity scores between books based on user-book interactions represented by the pivot table
similarity_scores = cosine_similarity(pivot_table)
similarity_scores

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

In [252]:
similarity_scores.shape

(706, 706)

#### Building a Function which recommends

In [253]:
# Find the index position of the book say "Animal Farm" in the pivot table
np.where(pivot_table.index=='Animal Farm')

(array([47], dtype=int64),)

In [170]:
index = np.where(pivot_table.index=='Animal Farm')[0][0]


In [171]:
sorted(similarity_scores[index], reverse =True)

[0.9999999999999998,
 0.2702651417103732,
 0.2416404696056939,
 0.2362570952024308,
 0.21897040971615078,
 0.20936355536567294,
 0.2039361513890485,
 0.2029392104193419,
 0.19936781793629338,
 0.1934411223332458,
 0.1933680785174527,
 0.1853698575738056,
 0.18432987234650272,
 0.1828523777485272,
 0.1808593793358965,
 0.18033685629548887,
 0.17857624365011393,
 0.17810293434375038,
 0.17574683842508246,
 0.17550919073101653,
 0.17524780797774,
 0.17285553764932207,
 0.17139647009659337,
 0.1687005636608251,
 0.16727198067474489,
 0.16082459779343755,
 0.16077267549550414,
 0.15870304913591407,
 0.15866348262730293,
 0.1565007975327465,
 0.15632476361389702,
 0.1548168121432704,
 0.1526938495910763,
 0.15260868572873312,
 0.15109390524068056,
 0.1505511478256334,
 0.15009211603421194,
 0.1486861806511982,
 0.14731534531412904,
 0.1472959089266509,
 0.1470104920254031,
 0.14453100696863436,
 0.14436961311545574,
 0.1441287267186271,
 0.1432521990496692,
 0.1426711909607604,
 0.1412494068

In [254]:
# Sort the enumerated cosine similarity scores for the book "Animal Farm" in descending order
# without enumerate, we we only have similarity_score and we lost the index information to fetch boock information
sorted(list(enumerate(similarity_scores[index])),key=lambda x:x[1],reverse=True)

[(47, 0.9999999999999998),
 (0, 0.2702651417103732),
 (45, 0.2416404696056939),
 (324, 0.2362570952024308),
 (422, 0.21897040971615078),
 (92, 0.20936355536567294),
 (344, 0.2039361513890485),
 (309, 0.2029392104193419),
 (500, 0.19936781793629338),
 (82, 0.1934411223332458),
 (695, 0.1933680785174527),
 (86, 0.1853698575738056),
 (648, 0.18432987234650272),
 (650, 0.1828523777485272),
 (238, 0.1808593793358965),
 (389, 0.18033685629548887),
 (304, 0.17857624365011393),
 (175, 0.17810293434375038),
 (40, 0.17574683842508246),
 (178, 0.17550919073101653),
 (432, 0.17524780797774),
 (610, 0.17285553764932207),
 (19, 0.17139647009659337),
 (192, 0.1687005636608251),
 (507, 0.16727198067474489),
 (156, 0.16082459779343755),
 (202, 0.16077267549550414),
 (691, 0.15870304913591407),
 (5, 0.15866348262730293),
 (312, 0.1565007975327465),
 (667, 0.15632476361389702),
 (626, 0.1548168121432704),
 (318, 0.1526938495910763),
 (384, 0.15260868572873312),
 (550, 0.15109390524068056),
 (258, 0.15055

In [255]:
def recommend(book_name):
    """
    Recommends similar books based on the input book name.

    Parameters:
    - book_name (str): The name of the book for which recommendations are sought.

    Returns:
    - data (list): A list containing information about recommended books, including title, author, publication year, and image URL.
    """

    # Find the index position of the input book in the pivot table
    index = np.where(pivot_table.index == book_name)[0][0]

    # Find similar items based on cosine similarity scores, excluding the input book itself
    similar_items = sorted(list(enumerate(similarity_scores[index])), key=lambda x: x[1], reverse=True)[1:5]

    # Prepare data for recommended books
    data = []
    for i in similar_items:
        item = []
        # Extract information about recommended books from the books dataset
        temp_df = books[books['Book-Title'] == pivot_table.index[i[0]]]
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Title'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Author'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Year-Of-Publication'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Image-URL-M'].values))
        
        data.append(item)
    
    return data

In [256]:
# Checking and verifying the recommendations for a Book
recommend('Animal Farm')

[['1984',
  'George Orwell',
  1990,
  'https://images.amazon.com/images/P/0451524934.01.MZZZZZZZ.jpg'],
 ['Angus, Thongs and Full-Frontal Snogging: Confessions of Georgia Nicolson',
  'Louise Rennison',
  2001,
  'https://images.amazon.com/images/P/0064472272.01.MZZZZZZZ.jpg'],
 ['Midnight',
  'Dean R. Koontz',
  1995,
  'https://images.amazon.com/images/P/0425118703.01.MZZZZZZZ.jpg'],
 ['Second Nature',
  'Alice Hoffman',
  1994,
  'https://images.amazon.com/images/P/0399139087.01.MZZZZZZZ.jpg']]

In [258]:
# List of books there in the pivot table
list(pivot_table.index)

['1984',
 '1st to Die: A Novel',
 '2nd Chance',
 '4 Blondes',
 'A Bend in the Road',
 'A Case of Need',
 'A Child Called \\It\\": One Child\'s Courage to Survive"',
 'A Civil Action',
 'A Day Late and a Dollar Short',
 'A Fine Balance',
 'A Great Deliverance',
 'A Heartbreaking Work of Staggering Genius',
 'A Is for Alibi (Kinsey Millhone Mysteries (Paperback))',
 'A Lesson Before Dying (Vintage Contemporaries (Paperback))',
 'A Man Named Dave: A Story of Triumph and Forgiveness',
 'A Man in Full',
 'A Map of the World',
 'A Painted House',
 'A Patchwork Planet',
 'A Prayer for Owen Meany',
 'A Thin Dark Line (Mysteries &amp; Horror)',
 "A Thousand Acres (Ballantine Reader's Circle)",
 'A Time to Kill',
 "A Virtuous Woman (Oprah's Book Club (Paperback))",
 'A Walk to Remember',
 'A Widow for One Year',
 'A Wrinkle In Time',
 'A Wrinkle in Time',
 'A Year in Provence',
 "ANGELA'S ASHES",
 'About a Boy',
 'Absolute Power',
 'Acceptable Risk',
 'Accident',
 'Airframe',
 'All Around the To

In [176]:
# Replace 'http' with 'https' in the 'Image-URL-M' column
books['Image-URL-M'] = books['Image-URL-M'].str.replace('http://', 'https://')

# Verify the changes
books['Image-URL-M'].head()

0    https://images.amazon.com/images/P/0195153448....
1    https://images.amazon.com/images/P/0002005018....
2    https://images.amazon.com/images/P/0060973129....
3    https://images.amazon.com/images/P/0374157065....
4    https://images.amazon.com/images/P/0393045218....
Name: Image-URL-M, dtype: object

#### Serialize (Storing) the impoartant dataframes(informations) For buliding WebApp using pickle


In [177]:
import pickle

In [260]:
# Remove duplicate entries in the books DataFrame based on the 'Book-Title' column
books = books.drop_duplicates('Book-Title')

books

ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271354  0449906736      Flashpoints: Promise and Peril in a New World   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271354          Robin Wright                1993   
271356            Teri Sloat                1991   
271357      Christine Wicker                2004   
271358                 Plato                1996   
271359   Christopher  Biffle                2000   

                                               Publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                             W. W. Norton &amp; Company   
...                                                  ...   
271354                                  Ballantine Books   
271356                                      Dutton Books   
271357                                HarperSanFrancisco   
271358                           Oxford University Press   
271359  McGraw-Hill Humanities/Social Sciences/Languages   

                                              Image-URL-S  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271354  http://images.amazon.com/images/P/0449906736.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-M  \
0       https://images.amazon.com/images/P/0195153448....   
1       https://images.amazon.com/images/P/0002005018....   
2       https://images.amazon.com/images/P/0060973129....   
3       https://images.amazon.com/images/P/0374157065....   
4       https://images.amazon.com/images/P/0393045218....   
...                                                   ...   
271354  https://images.amazon.com/images/P/0449906736....   
271356  https://images.amazon.com/images/P/0525447644....   
271357  https://images.amazon.com/images/P/006008667X....   
271358  https://images.amazon.com/images/P/0192126040....   
271359  https://images.amazon.com/images/P/0767409752....   

                                              Image-URL-L  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.ama

In [261]:
# Checking the Count of duplicated entries in the books DataFrame based on the 'Book-Title' column
books.duplicated('Book-Title').sum()

0

In [264]:
# Serializing (Store) the popular_df, pivot table, books DataFrame, and similarity scores using pickle

# Dump the popular_df DataFrame into a binary file named 'popular.pkl'
pickle.dump(popular_df,open('popular.pkl','wb'))

# Dump the pivot table into a binary file named 'pt.pkl'
pickle.dump(pivot_table, open('pivot_table.pkl', 'wb'))

# Dump the books DataFrame into a binary file named 'books.pkl'
pickle.dump(books, open('books.pkl', 'wb'))

# Dump the similarity scores into a binary file named 'similarity_scores.pkl'
pickle.dump(similarity_scores, open('similarity_scores.pkl', 'wb'))